Improve `NcCheckboxRadioSwitch` and use for `NcAppSidebarTabs`

susnux opened the pull request:
* Resolves #603

I split this into separate commits to make reviewing easier as the PR is quite huge.
The first part adjusts the styling of the checkbox radio switch to be used with for the sidebar tabs (adjust borders and add an icon slot).

## Part 1: Improve `NcCheckboxRadioSwitch` styling

### Before
The hover colors are confusing, primary color is applied to multiple elements, the border is only applied to not-selected elements.

https://user-images.githubusercontent.com/1855448/229255077-560d4054-8cbf-4207-a4e5-4902bf887f8d.mp4

<details>

### With improvements (added borders)
The border is applied around the whole group, the hover color is fixed to be primary on the checked element and default hover color otherwise.

https://user-images.githubusercontent.com/1855448/229255140-24e1603a-c0cf-436c-83f8-053bfbe85533.mp4

</details>

### With improvements
The border is applied around the whole group, the hover color is fixed to be primary on the checked element and default hover color otherwise.
Added rounded borders (last commit) to match the mockup in the linked issue. 

https://user-images.githubusercontent.com/1855448/229257472-4cc21a34-1956-414e-b872-4c2ce742f709.mp4

## Part 2: Use `NcCheckboxRadioSwitch` for `NcAppSidebarTabs`

before | after
---|---
![image](https://user-images.githubusercontent.com/1855448/229257972-678ee373-3d5a-4f7a-a220-4a1a945ad20b.png)|![image](https://user-images.githubusercontent.com/1855448/229499855-d34cb3d2-6adc-4931-b00b-399b1848edbc.png)

Outdated initial version:
<details>

before | after
---|---
![image](https://user-images.githubusercontent.com/1855448/229257972-678ee373-3d5a-4f7a-a220-4a1a945ad20b.png)|![image](https://user-images.githubusercontent.com/1855448/229257980-e7077974-aa9c-4abe-973d-3ba453ae290d.png)

https://user-images.githubusercontent.com/1855448/229258351-8b5406c9-33ec-4eab-94eb-2de81def30f4.mp4

</details>

jancborchardt reviewed (commented):
Looks much much better!

@susnux a question: On part 1 "with improvements, should the single-line radio buttons not have more border-radius so they are pill-style? (Like we do for the View switcher in Forms currently.)

susnux commented:
@jancborchardt 
> a question: On part 1 "with improvements, should the single-line radio buttons not have more border-radius so they are pill-style? (Like we do for the View switcher in Forms currently.)

I used your and @nimishavijay comment here as a reference for how it should look like: https://github.com/nextcloud/nextcloud-vue/issues/603#issuecomment-1236878789

jancborchardt commented:
@susnux ah thank you for the explanation! :) I would just say let’s give it a bit more border-radius so that when it’s just one line it ends up being pill-style like all our buttons (and the current switcher style in Forms), as otherwise it will look out of place?

The radius should be 22px probably, or `calc(var(--default-clickable-area) / 2)`. Then it should be nicely pill-style for single-line buttons and also work well for higher ones like in the sidebar.

susnux commented:
@jancborchardt 
> I would just say let’s give it a bit more border-radius so that when it’s just one line it ends up being pill-style like all our buttons (and the current switcher style in Forms), as otherwise it will look out of place?

I changed that, it does now look like this:
![pill like buttons](https://user-images.githubusercontent.com/1855448/229499556-3d0ce3f9-c785-45e5-9670-ea166fb58602.png)

raimund-schluessler commented:
> @jancborchardt
> 
> > I would just say let’s give it a bit more border-radius so that when it’s just one line it ends up being pill-style like all our buttons (and the current switcher style in Forms), as otherwise it will look out of place?
> 
> I changed that, it does now look like this: ![pill like buttons](https://user-images.githubusercontent.com/1855448/229499556-3d0ce3f9-c785-45e5-9670-ea166fb58602.png)

I think that is to much. `calc(var(--default-clickable-area) / 2) i.e. 22px` works better, I think. And the outer element needs a radius `2px` larger than the inner one, to account for the border-thickness. Otherwise, there will be gaps at the corners.

jancborchardt commented on src/components/NcCheckboxRadioSwitch/NcCheckboxRadioSwitch.vue:
```diff
@@ -492,49 +540,83 @@ $spacing: 4px;
+		--border-radius-nc-radio-switch: var(--border-radius-large);
+
+		&-h-grouped {
+			--border-radius-nc-radio-switch: var(--border-radius-pill);
```
Here I would say to use `calc(var(--default-clickable-area) / 2)` instead.

This will result in a consistent border-radius of 22px.
- Then it looks pill-style for single-line switches like in Forms.
- And for double-line (like with icon above text like in the Files sidebar) it will also look nice and not have a huge radius


@susnux Hope that clears it up, sorry for the confusion. :)

jancborchardt replied:
Ah, just saw that @raimund-schluessler already commented the same – that's exactly what I meant, thanks! :)

susnux replied:
> Then it looks pill-style for single-line switches like in Forms.

BTW we use `--border-radius-pill` on forms too, so this might look a bit different when using this.

raimund-schluessler commented on src/components/NcCheckboxRadioSwitch/NcCheckboxRadioSwitch.vue:
```diff
@@ -545,17 +627,20 @@ $spacing: 4px;
+		&:not(:first-of-type) {
+			border-left: 0!important;
 		}
 		& + &.checkbox-radio-switch--checked {
```
I think this rule needs to go. It makes the center tab move when activating the right one.

raimund-schluessler replied:
> The rule below was the problem, as it is now longer needed.

Yeah, I just couldn't comment there as it was not edited in this PR 🙈

raimund-schluessler commented on src/components/NcCheckboxRadioSwitch/NcCheckboxRadioSwitch.vue:
```diff
@@ -545,17 +627,20 @@ $spacing: 4px;
 		&:first-of-type {
-			border-top-left-radius: var(--border-radius-large);
-			border-bottom-left-radius: var(--border-radius-large);
+			border-top-left-radius: var(--border-radius-nc-radio-switch);
```
These values need to be `2px` larger than the ones for the inner buttons to account for the border-thickness.

raimund-schluessler replied:
Just for reference, here you see this gap:
![image](https://user-images.githubusercontent.com/2496460/229503691-505dfdca-397e-4808-9f01-95bee9d81331.png)

susnux replied:
:+1:

susnux commented:
@jancborchardt @raimund-schluessler Do we want the same styling on the vertical group? That would look like this:

![image](https://user-images.githubusercontent.com/1855448/229512155-adc1c220-ec95-4bbf-9c4f-0fb4dbe205e2.png)

Or do we want to keep the previous design for the vertical buttons?

![image](https://user-images.githubusercontent.com/1855448/229512303-60181bb7-0058-41cd-9527-8a84b0f6bdb4.png)

susnux commented:
Ok I fixed the outer border, used the radius for both vertical and horizontal button groups and fixed the inner border margin. I think it is now ready to go.

It looks like this:

https://user-images.githubusercontent.com/1855448/229516798-44b4b9a8-6310-4937-9536-68507c665931.mp4

jancborchardt reviewed (approved):
Good stuff, let's go with that! :)

raimund-schluessler commented:
@susnux works nicely now in the docs. I will test it in a real app tomorrow.
One of the unit tests still needs a fix, though.

skjnldsv commented:
![Capture d’écran_2023-04-04_08-18-04](https://user-images.githubusercontent.com/14975046/229704481-f3250e8b-551d-4e11-b351-12edfa2d9796.png)

I find it a bit weird to have those pill rounded borders next to each others :)
I didn't thought that much into it, but couldn't we have straight borders when there is another button next to it and leave the rounded pill for the first and last items?
![image](https://user-images.githubusercontent.com/14975046/229704998-15a7bcd3-1202-45d4-a822-983858aee6e1.png)

I'll let the design team take the decision, it's juste a thought :)

skjnldsv reviewed (approved):
Besides one comment, looks really good! Nice job @susnux !

marcoambrosini reviewed (commented):
Hi there, design-wise I would suggest to:

- loose the border
- add a 4px gap between the buttons
- increase the vertical padding around the labels 

https://user-images.githubusercontent.com/26852655/229856459-f4f52857-c92a-4673-bb08-519be7046f0a.mov

susnux commented:
I agree on the gap, I think it looks better :)

But I personally do not like the missing border, it feels like something is missing visually.

nimishavijay commented:
Agreed with @susnux that a border looks nice! We should also keep in mind that in some apps there are more than 3 sidebar tabs which would look fairly crowded on small screens (this is a bigger issue, but that's another discussion altogether). So I suggest that we keep the spacing to 2px. What do you think @marcoambrosini ?

jancborchardt commented:
I think the improved vertical spacing is good @marcoambrosini!

Agree with @susnux the border is needed, especially to cover cases where there are elements other than the radio group horizontally or vertically. E.g. in Forms there is the Share button and sidebar toggle next to it.

Regarding the gap I'm not sure mostly for space reasons. But I'd be fine with anything that's decided. :)

susnux commented:
Also there is no space for padding, the tabs have to fit into 300px to 500px (our current limits of the sidebar menu).

For a "real" app you will get following on the minimum size and three tabs:

2px gap | 4px gap
---|---
![screenshot](https://user-images.githubusercontent.com/1855448/230146291-d5fa55ad-442b-4176-9599-2f4755d11e18.png)|![screenshot](https://user-images.githubusercontent.com/1855448/230146267-3a488201-77b1-4a7b-be93-8d65bc5e34e6.png)

So I would go along with @nimishavijay on that 2px.

jancborchardt reviewed (approved):
Looks great now I think! :)

raimund-schluessler reviewed (approved):
Works well with the Tasks app. The failing test still needs a fix, but I have no idea, why it fails, honestly. Seems the `NcCheckboxRadioSwitch` does not emit the `onToggle` event in the tests, but it works well in a real app.

raimund-schluessler commented on tests/unit/components/NcAppSidebar/NcAppSidebarTabs.spec.js:
```diff
@@ -115,22 +115,22 @@ describe('NcAppSidebarTabs.vue', () => {
 			it('emit "update:active" event with the selected tab id when clicking on a tab', () => {
-				const lastLink = wrapper.find('nav>ul>li:last-of-type>a')
+				const lastLink = wrapper.find('nav>.checkbox-radio-switch:last-of-type>label')
 				lastLink.trigger('click')
```
> Maybe `await`ing the trigger fixes the test?

No, it does not :see_no_evil:

ShGKme replied:
More info: a click on the `<label>` in `<NcCheckboxRadioSwitch>` triggers a `click` on `<input>`, but does not trigger a `change` event.

Seems to be either jsdom or VTU bug, but all related bug reports are fixed.

*continues investigating*

nimishavijay commented:
Well, I believe that looking like radio buttons is kind of the idea! :) @ShGKme 

Adding a border makes sure that:
- The items are visually grouped together
- It doesn't look too similar to a button, which is usually used to do some action. This component will be used to switch between multiple views
- The alignment of the item is clear. Without the border, it is difficult to determine where the leftmost item begins, thereby making it look like the item is not aligned neatly with the text above or below it (for example, in Marco's video, it seems like "Search" is not neatly aligned with `cat-picture.jpg / last edited 3 weeks ago`, and adding a border would help)

susnux commented:
Why I prefer the current style (borders and all round):
* It is consistent with the other styles (we always have rounded elements
* The border groups belonging elements
* It makes clear where the element ends, so you know where to click (I think this is an important point regarding accessibility

raimund-schluessler commented:
@jancborchardt I think we need a design decision here, and someone who feels confident to press "Merge" :wink:

jancborchardt commented:
I agree with the points brought up especially regarding accessibility and grouping.

For simplicity, I suggest:
- We go with this design now and merge it, rounded and border around, as it is much better than what is there now and in line with the rest of our design
- If we want to adjust its occurrence in the right sidebar navigation (as tabs) to use no border, we can very simply do that.

So I’ll merge it to unblock the situation. :v:
